### PR TITLE
chore(runner-ct): TS and React improvements

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes N/A

### User facing changelog

N/A

### Additional details
 
Improve TypeScript and React code in RunnerCt.

Note: This carries on from the work in https://github.com/cypress-io/cypress/pull/16613. It it targeting that branch, not develop.

This is just a refactor. I changed the class component into a function component, and improved the types (slightly, there are still some untyped arguments, where it's not obvious what they types should be). I also eliminated some unused code.

### How has the user experience changed?

No change.

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has the original issue or this PR been tagged with a release in ZenHub? 
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/runner-ct/src/header/header.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/header/header.tsx:16]: This type doesn't seem to exist anywhere. We should try to enable `strict` asap.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/iframe/iframes.tsx:95]: The type of `snapshotProps` is a mystery. I am not sure if the type exists in the code base. So I just left it for now.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/iframe/iframes.tsx:96]: Obviously mutation `props.state` is not very good - I tried to make this refactor without changing the logic at all. I'll fix this in another pass in the future.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/iframe/iframes.tsx:121]: Another mutation of props. Again - just using the code as-is, trying to minimize changes.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/iframe/iframes.tsx:95]: I don't know why it's complaining here.

Also, I think I'm supposed to pass this as a separate prop that is wrapped in a `useCallback`. Is this correct? @agg23

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [inline packages/runner-ct/src/iframe/iframes.tsx:89]: I want this to be like `componentDidMount` (that how's the code worked, pre-refactor). My feeling is passing all the require args in the array is not going to do this, though. You can remove the eslint disable and run `yarn lint` and see what I mean.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:78]: `disposers` isn't actually used outside of this hook. You're just using it to represent all of the things you need to clean up. Just assign this array to a local var and call `dispose` on them in the cleanup method (as you're already doing).

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [reply packages/runner-ct/src/iframe/iframes.tsx:89]: Yes. You have it written correctly. The ESLint rule doesn't cover a standard `componentDidMount` scenario.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:32]: This is entirely local to the `useEffect`

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:27]: I know you don't like destructuring props, but you really should do it. It would clean up this code quite a bit.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [reply packages/runner-ct/src/iframe/iframes.tsx:95]: It's complaining because it doesn't like the deep property access. It would stop if you assigned `props.state.callbackAfterUpdate` to a variable.

I don't understand your question.

Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:95]: This is not a correct translation from the class component. `componentDidUpdate`  is called on any prop or state change, except on initial mount. As written, your method will run exactly once assuming the reference `componentDidUpdate` never changes.

To copy exactly, you want something like:
```typescript
const isFirstRenderRef = useRef(true)

useEffect(() => {
  if (isFirstRenderRef.current) {
    isFirstRenderRef.current = false
    return
  }

  props.state.callbackAfterUpdate?.()
})
```

Notice that the `useEffect` does not provide a dependency array. This causes it to run on _every_ render.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [reply packages/runner-ct/src/iframe/iframes.tsx:97]: Yea, I was running into this as well. You could make a simple incomplete type, as we know it at least has the form:

```typescript
interface SnapshotProps {
  snapshots: Array<{ body: unknown }>
  $el?: unknown // Probably HTMLElement or Jquery.Element
}
```

But I'll leave that up to you to decide.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [reply packages/runner-ct/src/iframe/iframes.tsx:98]: I expect this is actually correct. `state` is likely a MobX store, so there's nothing wrong with mutating it. That being said, it's much nicer to write methods on the store to perform the mutation, rather than doing it yourself inline.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:50]: nit: Effects should be the last thing in the component before any JSX logic, since they're the last thing to run. That also ensure you've actually defined all of the variables you've referenced when the lambda is created (`_setScriptError` doesn't get defined until later in the component at the moment)

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:147]: You might have to handle `containerRef.current === null` here

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [inline packages/runner-ct/src/iframe/iframes.tsx:51]: Ideally I think all of these would be registered from within `AutIframe`

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (agg23) [reply packages/runner-ct/src/iframe/iframes.tsx:78]: Using `setState` in this way is forcing the component to rerender _AND_ doesn't even work correctly:

Initial render -> run effect + setState -> render again

After some time

`componentWillUnmount` -> run effect cleanup -> run `disposers.forEach` on the reference of `disposers` that existed _when the hook first ran_. Since the `useEffect` has not run since the initial render, `disposers` is an empty array, and nothing gets disposed.

----

It's important to remember that the dependency arrays in React hooks are not simply "rerun when one of these changes". They define what values in the closure are allowed to be updated, so if a value doesn't appear in the dependency array, that value never changes in the hook.

**Review by agg23 — DISMISSED** [review]

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:98]: I see. Technically correct or not, I agree with you, it *looks* weird and just seems unidiomatic. I'll make a method to handle the mutation.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:95]: I see - this warning is basically a false positive, then. This pattern (deep property access) is used in the React docs:

image

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:27]: We do get the same type safety either way, but happy to make this change. It seems a bit more concise.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:29]: Looks like `useRef` is what I wanted. I get why it's `useRef` - it's a reference to a value, not an actual value. I wish this was named `useMutableValue`, or something more explicit, though.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:78]: Right, this makes perfect sense. I am not sure what I was thinking with `useState` here.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:27]: Sure. I am not entirely sure how destructing makes this cleaner, though. Are you suggesting just to do something like:

```js
export const Iframes: React.FC<IFramesProps> = namedObserver('Iframes', ({
  state,
  eventManager
  // ... etc ...
}) => { 
```

?

Happy to make this change, I'd like to know what makes this cleaner/better than the existing style, though - I've seen the word "convention" tossed around, but no-one seems to be able to explain why it's objectively superior.

I don't have a strong opinion about this, it's just a coding style, but I don't want to do something without understanding why it's an improvement.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:50]: Makes sense - I made this change.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:147]: I just added a defensive check. Not sure if it can even be null at the time this code runs, but better safe than sorry.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:51]: I think so, too. That said, since this PR is already making a lot of changes to a fairly important part of the code, and moving them to `AutIframe` would be more than just a simple clean up (basic goal of this PR), I'd prefer to do this later (if we decide to do it at all).

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx]: Fixed.

File: packages/runner-ct/src/iframe/iframes.tsx
Comment (lmiller1990) [reply packages/runner-ct/src/iframe/iframes.tsx:35]: I'll spend some time figuring out the entire type at some point. I kind of feel like these partially typed objects do more harm than good - it's often not clear a type is only partially complete, and the reads incorrectly assumes it is.

**Comment by elevatebart** [comment]
I believe you might have to create another PR since this one is blocked by Adam, but this is pretty good as is ;)

**Review by elevatebart — DISMISSED** [review]
I like that there is less mobx

**Review by elevatebart — APPROVED** [review]
Done

**Review by elevatebart — DISMISSED** [review]
Good?

**Comment by lmiller1990** [comment]
Not quite @elevatebart - there is a bunch of unrelated changes here now, since retargeting. 

Need to merge this first: https://github.com/cypress-io/cypress/pull/16612#pullrequestreview-667322321 and rebase. Please wait a bit :)

**Comment by lmiller1990** [comment]
Should be good for review now @elevatebart.

**Review by JessicaSachs — APPROVED** [review]

**Pull request merged**
